### PR TITLE
Code changes for PEAKS plugin

### LINKED ISSUE #800: K8s plugin for Power Efficiency Aware Kubernetes Scheduling (PEAKS)
### Area

- [X] Scheduler
- [ ] Controller
- [ ] Helm Chart
- [ ] Documents

### Other components

_No response_

### What happened?

The energy efficiency of different nodes of a Kubernetes cluster is not the same typically. However, current Kubernetes scheduler doesn't have any capability to distinguish the difference in the energy efficiency of different nodes to prioritize selection of the nodes with higher energy efficiency over nodes with lower energy efficiency for the pod placement. 

### What did you expect to happen?

Create a Kubernetes Enhancement Proposal (KEP) motivating the need for a scheduler plugin (to achieve energy efficiency aware scheduling of pods on nodes) and also capture the functional details of one such plugin. Subsequently implement the proposed energy efficiency aware K8s scheduler plugin.

### How can we reproduce it (as minimally and precisely as possible)?

_No response_

### Anything else we need to know?

_No response_

### Kubernetes version

<details>

```console
$ kubectl version
# paste output here
```

</details>

### Scheduler Plugins version

<details>

</details>

### PR DESCRIPTION
**Pull request by knarayan**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
Code changes for the scheduler plugin for energy efficiency aware scheduling of pods on nodes (prioritize nodes with higher energy efficiency over nodes with lower energy efficiency for pod placement)

#### Which issue(s) this PR fixes:

Fixes #800

#### Special notes for your reviewer:
NONE

#### Does this PR introduce a user-facing change?
NONE

```release-note

```

### CODE REVIEW

**Comment by knarayan** [comment]
This is the same PR as https://github.com/kubernetes-sigs/scheduler-plugins/pull/810 (fixing some branch issues in that PR)

**Comment by knarayan** [comment]
Hi @Huang-Wei: The integration test `peaks_test.go` is failing at line https://github.com/knarayan/scheduler-plugins/blob/master/test/integration/peaks_test.go#L236 with the error `not enough arguments in call` to the function [podScheduled](https://github.com/knarayan/scheduler-plugins/blob/master/test/integration/utils.go#L55) which is called in the same manner in other integration tests also (like https://github.com/knarayan/scheduler-plugins/blob/master/test/integration/targetloadpacking_test.go#L180).
I was also able to run the `peaks_test.go` on our local k8s cluster by running `make integration-test` from the `scheduler-plugins` directory.
Kindly let us know what the issue here.

**Comment by Huang-Wei** [comment]
@knarayan try rebase the master - latest PRs get merged and changed the `podScheduled` to be:

```go
func podScheduled(t *testing.T, c clientset.Interface, podNamespace, podName string) bool {
```

**Comment by knarayan** [comment]
> @knarayan try rebase the master - latest PRs get merged and changed the `podScheduled` to be:
> 
> ```go
> func podScheduled(t *testing.T, c clientset.Interface, podNamespace, podName string) bool {
> ```

Thanks a lot, @Huang-Wei . Now the integration test `peaks_test.go` has passed.

File: pkg/trimaran/peaks/peaks.go [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+	for _, metric := range metrics {
+		if metric.Type == watcher.CPU {
+			if metric.Operator == watcher.Average || metric.Operator == watcher.Latest {
+				nodeCPUUtilPercent = metric.Value
+				cpuMetricFound = true
+			}
+		}
+	}
+	if !cpuMetricFound {
+		klog.ErrorS(nil, "Cpu metric not found in node metrics for nodeName", nodeName)
```
Comment (Huang-Wei) [inline pkg/trimaran/peaks/peaks.go:139]: Ditto. Eliminate klog by using contextual logging.

File: pkg/trimaran/peaks/peaks_test.go [hunk 2]
```diff
@@ -0,0 +1,657 @@
...
+
+func createSamplePowerModel() {
+	data := map[string]interface{}{
+		"node-1": map[string]float64{
+			"k0": 471.7412504314313,
+			"k1": -91.50493019588365,
+			"k2": -0.07186049052516228,
+		},
+	}
+	fmt.Println("Power model json data: ", data)
```
Comment (Huang-Wei) [inline pkg/trimaran/peaks/peaks_test.go:91]: do not use raw fmt.Print - pass-in a `testing.T` and use `t.sth()`

File: pkg/trimaran/peaks/peaks.go [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+	for _, model := range clusterPowerModel {
+		if maxPower < model.K0 {
+			maxPower = model.K0
+		}
+	}
+	return maxPower
+}
+
+func getPowerModel(nodeName string) PowerModel {
+	power_model, ok := clusterPowerModel[nodeName]
```
Comment (Huang-Wei) [inline pkg/trimaran/peaks/peaks.go:226]: Go uses camelCase naming

File: test/integration/peaks_test.go [hunk 4]
```diff
@@ -0,0 +1,247 @@
+/*
+Copyright 2021 The Kubernetes Authors.
```
Comment (Huang-Wei) [inline test/integration/peaks_test.go:2]: 2024

File: test/integration/peaks_test.go [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+	fwkruntime "k8s.io/kubernetes/pkg/scheduler/framework/runtime"
+	st "k8s.io/kubernetes/pkg/scheduler/testing"
+	imageutils "k8s.io/kubernetes/test/utils/image"
+
+	"sigs.k8s.io/scheduler-plugins/apis/config"
+	"sigs.k8s.io/scheduler-plugins/pkg/trimaran/peaks"
+	"sigs.k8s.io/scheduler-plugins/test/util"
+)
+
+func createSamplePowerModel() {
```
Comment (Huang-Wei) [inline test/integration/peaks_test.go:50]: could you:

- de-dupe createSamplePowerModel() as it's used in UT and integration test
- use defer() to remove temporary folder

File: pkg/trimaran/peaks/peaks_test.go [hunk 6]
```diff
@@ -0,0 +1,657 @@
+/*
+Copyright 2020 The Kubernetes Authors.
```
[the remark this answers is not in the document]
Comment (knarayan) [reply pkg/trimaran/peaks/peaks_test.go]: Done, @Huang-Wei !

File: test/integration/peaks_test.go (hunk 5 above)
Comment (knarayan) [reply test/integration/peaks_test.go]: Done @Huang-Wei !

File: test/integration/peaks_test.go (hunk 4 above)
Comment (knarayan) [reply test/integration/peaks_test.go]: Done, @Huang-Wei !

File: pkg/trimaran/peaks/peaks_test.go (hunk 2 above)
Comment (knarayan) [reply pkg/trimaran/peaks/peaks_test.go]: Done, @Huang-Wei !

File: pkg/trimaran/peaks/peaks.go (hunk 3 above)
Comment (fali007) [reply pkg/trimaran/peaks/peaks.go]: Made the change

File: pkg/trimaran/peaks/peaks.go (hunk 1 above)
Comment (fali007) [reply pkg/trimaran/peaks/peaks.go]: Changed to contextual logging.

File: pkg/trimaran/peaks/peaks.go [hunk 7]
```diff
@@ -0,0 +1,231 @@
...
+		return score, nil
+	}
+
+	var curPodCPUUsage int64
+	for _, container := range pod.Spec.Containers {
+		curPodCPUUsage += PredictUtilisation(&container)
+	}
+	if pod.Spec.Overhead != nil {
+		curPodCPUUsage += pod.Spec.Overhead.Cpu().MilliValue()
+	}
```
[the remark this answers is not in the document]
Comment (fali007) [reply pkg/trimaran/peaks/peaks.go]: Changed to use `resource.PodRequests`

File: pkg/trimaran/peaks/peaks.go [hunk 8]
```diff
@@ -0,0 +1,231 @@
...
+	if err = json.Unmarshal(data, &clusterPowerModel); err != nil {
+		klog.ErrorS(err, "Unable to unmarshal power model from the input configuration")
+		return err
+	}
+	return nil
+}
+
+func New(ctx context.Context, obj runtime.Object, handle framework.Handle) (framework.Plugin, error) {
+	logger := klog.FromContext(ctx)
+	klog.V(4).InfoS("Peaks plugin Input config %+v", obj)
```
[the remark this answers is not in the document]
Comment (fali007) [reply pkg/trimaran/peaks/peaks.go]: Made the change

File: pkg/trimaran/peaks/peaks.go [hunk 9]
```diff
@@ -0,0 +1,231 @@
...
+	return Name
+}
+
+func initNodePowerModels() error {
+	data, err := os.ReadFile(os.Getenv("NODE_POWER_MODEL"))
+	if err != nil {
+		klog.ErrorS(err, "Unable to read power model from the input configuration")
+		return err
+	}
+	if err = json.Unmarshal(data, &clusterPowerModel); err != nil {
```
[the remark this answers is not in the document]
Comment (fali007) [reply pkg/trimaran/peaks/peaks.go]: Changed to use strict decoding.

File: pkg/trimaran/peaks/peaks.go [hunk 10]
```diff
@@ -0,0 +1,231 @@
...
+var _ framework.ScorePlugin = &Peaks{}
+var maxPower = 0.0
+var clusterPowerModel map[string]PowerModel
+
+func (pl *Peaks) Name() string {
+	return Name
+}
+
+func initNodePowerModels() error {
+	data, err := os.ReadFile(os.Getenv("NODE_POWER_MODEL"))
```
[the remark this answers is not in the document]
Comment (fali007) [reply pkg/trimaran/peaks/peaks.go:66]: Made the change to get power model from peaksArgs. If peaksArgs is empty use env-variable to define read power model.

File: pkg/trimaran/peaks/peaks.go [hunk 11]
```diff
@@ -0,0 +1,231 @@
...
+	K0 float64 `json:"k0"`
+	K1 float64 `json:"k1"`
+	K2 float64 `json:"k2"`
+	// Power = K0 + K1 * e ^(K2 * x) : where x is utilisation
+	// Idle power of node will be K0 + K1
+}
+
+var _ framework.ScorePlugin = &Peaks{}
+var maxPower = 0.0
+var clusterPowerModel map[string]PowerModel
```
[the remark this answers is not in the document]
Comment (fali007) [reply pkg/trimaran/peaks/peaks.go]: Moved these inside Peaks Struct.

File: pkg/trimaran/peaks/deployment/deployment.yaml [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+        resources:
+          requests:
+            cpu: '0.1'
+        securityContext:
+          privileged: false
+        volumeMounts:
+        - mountPath: /shared
+          name: shared
+        - mountPath: /power_model
+          name: powermodel
```
[the remark this answers is not in the document]
Comment (fali007) [reply pkg/trimaran/peaks/deployment/deployment.yaml]: removed volume mount

File: pkg/trimaran/peaks/deployment/deployment.yaml [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+        command:
+        - /bin/kube-scheduler
+        - --bind-address=0.0.0.0
+        - --leader-elect=false
+        - --config=/home/scheduler-config.yaml
+        - -v=6
+        image: <REPLACE_ME_WITH_PEAKS_IMAGE>
+        env:
+        - name: NODE_POWER_MODEL
+          value: /power_model/node_power_model
```
[the remark this answers is not in the document]
Comment (fali007) [reply pkg/trimaran/peaks/deployment/deployment.yaml]: Removed env var definition.

File: pkg/trimaran/peaks/deployment/peaks-power-model-config.json [hunk 14]
```diff
@@ -0,0 +1 @@
+{"node_1":{"k0":"<replace with node 1 k0>","k1":"<replace with node 1 k1>","k2":"<replace with node 1 k2>"},"node_2":{"k0":"<replace with node 2 k0>","k1":"<replace with node 2 k1>","k2":"<replace with node 2 k2>"}}
```
[the remark this answers is not in the document]
Comment (knarayan) [reply pkg/trimaran/peaks/deployment/peaks-power-model-config.json:1]: We request to handle this separately through a different PR.

File: pkg/trimaran/peaks/deployment/README.md [hunk 15]
```diff
@@ -0,0 +1,65 @@
...
+Configuring power model
+
+```bash
+kubectl create configmap peaks-node-power-model --from-file=peaks-power-model-config.json -n kube-system
+```
+
+Deploy crds
+
+```bash
+kubectl apply -f peaks-crd.yaml
```
Comment (Huang-Wei) [inline pkg/trimaran/peaks/deployment/README.md:43]: where is peaks-crd.yaml?

File: pkg/trimaran/peaks/deployment/deployment.yaml [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+        securityContext:
+          privileged: false
+        volumeMounts:
+        - mountPath: /shared
+          name: shared
+      volumes:
+      - name: shared
+        hostPath:
+          path: /tmp
+          type: Directory
```
Comment (Huang-Wei) [inline pkg/trimaran/peaks/deployment/deployment.yaml:46]: I guess we don't need L37-L46?

File: apis/config/v1/types.go [hunk 17]
```diff
@@ -277,3 +277,22 @@ type SySchedArgs struct {
...
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// PeaksArgs holds arguments used to configure the Peaks plugin
+type PeaksArgs struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// Address of load watcher service
+	WatcherAddress string                `json:watcherAddress",inline"`
+	NodePowerModel map[string]PowerModel `json:nodePowerModel",inline"`
```
Comment (Huang-Wei) [inline apis/config/v1/types.go:289]: can you add a comment mentioning that you will be refactoring this in a follow-up PR? https://github.com/kubernetes-sigs/scheduler-plugins/pull/829#discussion_r1835548516

File: pkg/trimaran/peaks/peaks.go [hunk 18]
```diff
@@ -0,0 +1,214 @@
...
+	quantity := reqs[v1.ResourceCPU]
+	curPodCPUUsage := quantity.MilliValue()
+
+	var nodeCPUUtilPercent float64
+	var cpuMetricFound bool
+	for _, metric := range metrics {
+		if metric.Type == watcher.CPU {
+			if metric.Operator == watcher.Average || metric.Operator == watcher.Latest {
+				nodeCPUUtilPercent = metric.Value
+				cpuMetricFound = true
```
Comment (Huang-Wei) [inline pkg/trimaran/peaks/peaks.go:142]: add a `break` after L142?

File: pkg/trimaran/peaks/peaks_test.go [hunk 19]
```diff
@@ -0,0 +1,556 @@
...
+	"encoding/json"
+	"math"
+	"net/http"
+	"net/http/httptest"
+	"os"
+	"testing"
+
+	"github.com/paypal/load-watcher/pkg/watcher"
+	"github.com/stretchr/testify/assert"
+	testutil "sigs.k8s.io/scheduler-plugins/test/util"
```
Comment (Huang-Wei) [inline pkg/trimaran/peaks/peaks_test.go:30]: can you sort the imports into several sections and separate by empty lines?

- core Golang imports like `math`
- upstream imports like loadwatcher, k8s.io
- imports from this repo (sigs.k8s.io/scheduler-plugins)

File: pkg/trimaran/peaks/deployment/README.md (hunk 15 above)
Comment (fali007) [reply pkg/trimaran/peaks/deployment/README.md]: Changed this to correct filename

File: pkg/trimaran/peaks/deployment/deployment.yaml (hunk 16 above)
Comment (fali007) [reply pkg/trimaran/peaks/deployment/deployment.yaml]: Removed these lines

File: apis/config/v1/types.go (hunk 17 above)
Comment (fali007) [reply apis/config/v1/types.go:290]: Added comment on the followup PR feature

File: pkg/trimaran/peaks/peaks.go (hunk 18 above)
Comment (fali007) [reply pkg/trimaran/peaks/peaks.go:142]: Added break statement

File: pkg/trimaran/peaks/peaks_test.go (hunk 19 above)
Comment (fali007) [reply pkg/trimaran/peaks/peaks_test.go]: Rearranged the imports

**Comment by knarayan** [comment]
There are some branch issues in this PR. Hence we created a new PR https://github.com/kubernetes-sigs/scheduler-plugins/pull/841 and closing this.

File: pkg/trimaran/peaks/deployment/README.md [hunk 20]
```diff
@@ -0,0 +1,65 @@
...
+docker push <new image tag>
+```
+
+Configuring power model
+
+```bash
+kubectl create configmap peaks-node-power-model --from-file=peaks-power-model-config.json -n kube-system
+```
+
+Deploy crds
```
[the remark this answers is not in the document]
Comment (knarayan) [reply pkg/trimaran/peaks/deployment/README.md:40]: Commit https://github.com/kubernetes-sigs/scheduler-plugins/pull/841/commits/ccba9e89ff18c8aa5436c93a2d3e070c8fe8d6dc addresses this comment, @Huang-Wei .

**Pull request closed without merging**